keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,423 @@
+    - KUDO
+    - Kubernetes-workload
+
+    The output from diagnostics collection is an archive containing all
+    diagnostics artifacts.
+
+    ```bash
+    kubectl kudo diagnostics collect --instance=%instance% --namespace=%namespace%
+    ```
+
+3.  Diagnostics analysis
+
+    Is given an archive as input and provides a human-readable report as output
+    containing explanations for any identified issues.
+
+    ```bash
+    kubectl kudo diagnostics analyze cassandra_diagnostics.zip
+    ```
+
+### Operator developer experience
+
+To configure diagnostics globally, this KEP introduces an optional top-level
+`diagnostics` key in operator.yaml.
+
+#### Diagnostics collection
+
+The following diagnostics will be implicitly collected without any configuration
+from the operator developer:
+
+- Logs for deployed pods related to the KUDO Instance
+- YAML for created resources, including both spec and status, related
+  to the KUDO instance
+- Output of `kubectl describe` for all deployed resources related to the KUDO
+  Instance
+- Current plan status, if one exists, or the KUDO Instance
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
+- RBAC resources that are applicable to the KUDO controller manager
+- Current settings and version information for KUDO
+- Status of last preflight check run.
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle.resources`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - name: Zookeeper Configuration File
+        key: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: /opt/zookeeper/server.properties
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
```

Comment:
that would be nice, but we don't necessarily _know_ that at the moment, right? We'd be assuming a convention that doesn't really exist. I think it's better to ultimately have reference-able objects rather than magic strings or assumptions.

Reply:
tbf templating in operator.yaml makes me wretch too

Reply:
> I think it's better to ultimately have reference-able objects

I agree, but referencing a set of objects in k8s world is a solved problem using the `labelSelector`, right? You're assuming that all pods are within a `StatefulSet` but what if these pods are spread across multiple `ReplicaSet`s?  Maybe I'm missing something here.